### ISSUE TITLE
PR Template is not Helpful: What Would Work Better?

### ISSUE THREAD

**Issue by kevinkreiser** [valhalla/valhalla#3241]
A long while ago we added a PR template to the repo to encourage people to dot their T's and cross their I's before expecting an approving review and eventually merging their PR. Over time I have seen a great number of PRs sit with that unmodified template in their description. On occasion someone will fill out the description but leave the checklist unmodified. The vast majority of people delete the template and roll their own description.

I think the main goals of the pr template are not being served by its current form. What are the goals exactly? I think there are two big ones:

1. **Help the submitter know what they should do.** Things like: linting, unit test coverage, regression tests, performance tests, design discussion, documentation changes, changelog, etc. The good news here is that we can automate most of these (except RAD, perf and design(this one could go in a separate linked issue presumably?)).
2. **Help the reviewer finish the review in more timely fashion.** We do this mainly by passing all the automatic checks and then providing enough information about the change so that the reviewer doesnt have to reverse engineer the code to determine what is going on when its non-trivial.

I am not sure what the optimal solution to these problems is but I have a couple of ideas..

* remove the pr template or reduce it to some all caps warning and rely on CI checks instead, hear me out..
* if we automate as much as possible, adding a github action for pretty much every one of these light checks and we name them intuitively, i think people will be able to tell what is wrong with their pr
* as mentioned above we cant do RAD (yet) nor is perf something we can do (reliably) in CI. design discussions are out of scope for automation as well, but reviewers can open that can of worms when needed if a pr comes without an issue. basically it will be the submitters and reviewers discretion when any of these checks are warranted
* add a check that makes sure the submitter actually added a description to the pr that was non trivial

**Comment by nilsnolde** [comment]
Agreed, a more wholesome CI can definitely help preventing (first) submitter's frustration, e.g. pulling out the format.sh check into its own action (as you did with changelog) and let that fail all other workflows.

A few bullets/sections are still good for guiding a PR submission I think, again e.g. format.sh (probably the most popular commit message;)). 

From my pov and the past non-trivial PRs: it can become pretty unreadable and hard to figure out what was done why. IMO it's both the submitter's and the main maintainers' (plural) responsibility to ensure the main discussion on design/general implementation happens in the issue before working on the PR. Can save both reviewing and development time. But it's upfront work for both and needs more experienced people to take the lead so it doesn't take weeks to finish the discussion. Of course things can change and no one can foresee all details but I bet PRs like the GTFS one would probably have benefited from more conceptual beforehand discussions.

QGIS has smth called QEPs (Qgis Enhancement Proposals, like python's PEPs), totally different ball game in terms of code/review throughput, but it does really ensure the whole community can review the implementation plan and there's no miscommunication. Would probably be overshooting for this repository, but I think a thorough issue description for non-trivial PRs should be mandatory and no one should start working on a bigger PR without a maintainer's thumbs up. Ideally that would be guiding the implementation enough to save many refactorings and changes of heart during the PR submission. I'm also not very good at solving most things before working on a PR and mostly jump into it after rough initial approval (see #3008 , 99 comments going back and forth; really no hard feelings, just a good example where I should've fleshed it out more before).

I'd definitely be happy to discuss more beforehand, best including a lively discussion with the community and some more maintainers, if possible
